### PR TITLE
  PXB-2293 Merge MySQL 8.0.22

### PR DESCRIPTION
**Pull request by rahulmalik87**
PXB-2293 Merge MySQL 8.0.22
It  also contains the fix for 
PXB-2314 Handle Disk format changes introduced in MySQL 8.0.22 release

### CODE REVIEW

File: storage/innobase/log/log0recv.cc [hunk 1]
```diff
@@ -1684,19 +1652,17 @@ static byte *recv_parse_or_apply_log_rec_body(
...
-          !recv_sys->apply_file_operations));
 
     case MLOG_FILE_EXTEND:
 
       return (fil_tablespace_redo_extend(
           ptr, end_ptr, page_id_t(space_id, page_no), parsed_bytes,
-          !recv_sys->apply_file_operations));
->>>>>>> mysql-server/8.0
+          recv_sys->bytes_to_ignore_before_checkpoint != 0 ||
+              recv_sys->recovered_lsn + parsed_bytes <
```
Comment (rahulmalik87) [inline storage/innobase/log/log0recv.cc]: fix to address MLOG FILE EXTEND redo

**Comment by rahulmalik87** [comment]
ms 8.0.21 https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/183/

ms 8.0.22 https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/182/

File: storage/innobase/fil/fil0fil.cc [hunk 2]
```diff
@@ -10870,7 +10871,7 @@ byte *fil_tablespace_redo_extend(byte *ptr, const byte *end,
   }
 
   /* Offset within the file to start writing zeros */
-  os_offset_t offset = mach_read_from_8(ptr);
+  // os_offset_t offset = mach_read_from_8(ptr);
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: Why is this commented out? 
Not used ?

File: storage/innobase/fil/fil0fil.cc [hunk 3]
```diff
@@ -2803,22 +2803,23 @@ dberr_t Fil_shard::get_file_size(fil_node_t *file, bool read_only_mode) {
   }
 
   if (file->size == 0) {
+#ifndef XTRABACKUP
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: why was this moved a few lines above?

Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]:
```suggestion
#if !defined(UNIV_HOTBACKUP) && !defined(XTRABACKUP)
```

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 4]
```diff
@@ -2126,7 +2126,7 @@ bool copy_back(int argc, char **argv) {
       return (false);
     }
     auto key = Encryption::get_master_key_id();
-    int ret = fscanf(f, "%lu", &key);
+    int ret = fscanf(f, "%u", &key);
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: Ideally, this should come with a different commit, not squashed into the conflict resolution.

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 5]
```diff
@@ -2153,7 +2153,7 @@ bool copy_back(int argc, char **argv) {
 
     my_free(master_key);
 
-    ulint master_key_id;
+    uint32_t master_key_id;
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: Why we changed the type here?

File: storage/innobase/fil/fil0fil.cc [hunk 6]
```diff
@@ -10673,8 +10626,8 @@ byte *fil_tablespace_redo_create(byte *ptr, const byte *end,
 
 #else  /* !UNIV_HOTBACKUP */
 
-<<<<<<< HEAD
-  const auto files = fil_system->get_scanned_files(page_id.space());
+  const auto files =
+      fil_system->get_scanned_filename_by_space_id(page_id.space());
```
Comment (altmannmarcelo) [inline storage/innobase/fil/fil0fil.cc]: Why we changed from `get_scanned_files` to `get_scanned_filename_by_space_id` ?

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 7]
```diff
@@ -549,10 +549,6 @@ ENDIF()
 IF(WITHOUT_SERVER)
   MESSAGE(FATAL_ERROR "This branch does not support -DWITHOUT_SERVER=ON configuration.")
 ENDIF()
-# Disallow WITH_EMBEDDED_SERVER=OFF builds
-IF(NOT WITH_EMBEDDED_SERVER)
-  MESSAGE(FATAL_ERROR "This branch does not support -DWITH_EMBEDDED_SERVER=OFF configuration.")
-ENDIF()
```
Comment (satya-bodapati) [inline CMakeLists.txt:537]: Removed by upstream or you? Any reason to remove this?

File: storage/innobase/fil/fil0fil.cc (hunk 2 above)
Comment (rahulmalik87) [reply storage/innobase/fil/fil0fil.cc]: yes. it not used. was getting compilation error.

File: cmake/mysql_version.cmake [hunk 8]
```diff
@@ -38,11 +38,7 @@ CONFIGURE_FILE(
 
 MACRO(MYSQL_GET_CONFIG_VALUE keyword var file)
  IF(NOT ${var})
-<<<<<<< HEAD
    FILE (STRINGS ${CMAKE_SOURCE_DIR}/${file} str REGEX "^[ ]*${keyword}=")
```
Comment (satya-bodapati) [inline cmake/mysql_version.cmake]: shouldn't u use MYSQL_VERSION? btw, are we making the PXB version bump to server version in this release?

File: plugin/keyring/common/i_keyring_key.h [hunk 9]
```diff
@@ -62,6 +62,7 @@ struct IKey : public Keyring_alloc {
   virtual bool is_key_length_valid() = 0;
 
   virtual ~IKey() {}
+
```
Comment (satya-bodapati) [inline plugin/keyring/common/i_keyring_key.h]: remove

File: storage/innobase/fil/fil0fil.cc (hunk 3 above)
Comment (rahulmalik87) [reply storage/innobase/fil/fil0fil.cc]: it throwing compilation error extent_size being unused

File: plugin/keyring/common/keyring_key.h [hunk 10]
```diff
@@ -61,7 +61,7 @@ struct Key : IKey {
   bool is_key_length_valid() override;
 
  protected:
-  void set_key_type_enum(const std::string *key_type) override;
+  void set_key_type_enum(const std::string *key_type);
```
Comment (satya-bodapati) [inline plugin/keyring/common/keyring_key.h:64]: necessary to remove override? why?

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 5 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: Because of changes is get_master_key.

File: storage/innobase/ut/ut0dbg.cc [hunk 11]
```diff
@@ -56,15 +55,6 @@ void ut_set_assert_callback(std::function<void()> &callback) {
...
 #if !defined(UNIV_HOTBACKUP) && !defined(UNIV_NO_ERR_MSGS) && \
     !defined(XTRABACKUP)
-=======
-/** Report a failed assertion.
-@param[in] expr The failed assertion
-@param[in] file Source file containing the assertion
-@param[in] line Line number of the assertion */
-[[noreturn]] void ut_dbg_assertion_failed(const char *expr, const char *file,
-                                          ulint line) {
-#if !defined(UNIV_HOTBACKUP) && !defined(UNIV_NO_ERR_MSGS)
```
Comment (satya-bodapati) [inline storage/innobase/ut/ut0dbg.cc]: I think these conflicts are related to oxygen style changes. Better to include them

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 4 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc]: it throwing compilation because of return type changed of get_master_key_id()

File: storage/ndb/mcc/frontend/dojo/dijit/package.json [hunk 12]
```diff
@@ -6,11 +6,7 @@
 	},
 	"main": "main",
 	"dependencies": {
-<<<<<<< HEAD
 		"dojo":"1.16.2"
-=======
-		"dojo": "1.15.4"
->>>>>>> mysql-server/8.0
 	},
```
Comment (satya-bodapati) [inline storage/ndb/mcc/frontend/dojo/dijit/package.json]: since we don't care about ndb, shouldn't we take upstream version changes. ie. 1.15.4 ? wait did they lower version requirements?
HEAD is our source and mysql-sever/8.0 is upstream 8.0 right? :thinking:

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/ut/ut0ut.cc [hunk 13]
```diff
@@ -519,15 +519,10 @@ const char *ut_strerr(dberr_t num) {
       return ("Sample reader has been requested to stop sampling");
     case DB_OUT_OF_RESOURCES:
       return ("System has run out of resources");
-<<<<<<< HEAD
-
     case DB_PAGE_IS_BLANK:
```
Comment (satya-bodapati) [inline storage/innobase/ut/ut0ut.cc]: This is used by PXB?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 7 above)
Comment (rahulmalik87) [reply CMakeLists.txt:537]: removed by upstream

File: plugin/keyring/common/keyring_key.h (hunk 10 above)
Comment (satya-bodapati) [reply plugin/keyring/common/keyring_key.h:64]: ignore

**Comment by altmannmarcelo** [comment]
@rahulmalik87 also can you please make sure PR and commits follow Percona Standard (PXB ISSUE, link, ...) and make sure you have individual JIRA Tickets for specific issues (such as Archiver log).

File: storage/innobase/fil/fil0fil.cc (hunk 6 above)
Comment (rahulmalik87) [reply storage/innobase/fil/fil0fil.cc]: ms 8.0.22 change get_scanned_files to get_scanned_filename_by_space_id

File: storage/ndb/mcc/frontend/dojo/dijit/package.json (hunk 12 above)
Comment (rahulmalik87) [reply storage/ndb/mcc/frontend/dojo/dijit/package.json]: I got the recommendation from GitHub to update the version of the dojo as a security alert. It was done in the previous release. so should do the changes according to upstream or let it be 1.16.2 to avoid security alert of the repo

**Comment by rahulmalik87** [comment]
> @rahulmalik87 also can you please make sure PR and commits follow Percona Standard (PXB ISSUE, link, ...) and make sure you have individual JIRA Tickets for specific issues (such as Archiver log).

Hey @altmannmarcelo I have changed the commits as part of this merge.

File: storage/ndb/mcc/frontend/dojo/dijit/package.json (hunk 12 above)
Comment (satya-bodapati) [reply storage/ndb/mcc/frontend/dojo/dijit/package.json]: I think we leave it same as upstream. check with kamil.

File: storage/innobase/ut/ut0ut.cc (hunk 13 above)
Comment (rahulmalik87) [reply storage/innobase/ut/ut0ut.cc]: yes  case DB_PAGE_IS_BLANK:
        return ("Page is blank");
. it was in the previous release as well.

File: cmake/mysql_version.cmake (hunk 8 above)
Comment (rahulmalik87) [reply cmake/mysql_version.cmake]: ${file} we were using in the previous release instead of using VERSION file. yes, we are bumping PXB version to use 8.0.22 instead of 8.0.15.

File: storage/ndb/mcc/frontend/dojo/dijit/package.json (hunk 12 above)
Comment (rahulmalik87) [reply storage/ndb/mcc/frontend/dojo/dijit/package.json]: leaving it as upstream. will check with team if GitHub complains again.

File: cmake/mysql_version.cmake (hunk 8 above)
Comment (satya-bodapati) [reply cmake/mysql_version.cmake]: The conflict resolution shows that you still use {file}. Shouldn't you use `${CMAKE_SOURCE_DIR}/MYSQL_VERSION}` instead of `${file}`

File: sql/CMakeLists.txt [hunk 14]
```diff
@@ -566,11 +566,8 @@ SET(SQL_SHARED_SOURCES
   tztime.cc
   uniques.cc
   xa.cc
-<<<<<<< HEAD
-  ../libmysql/libmysql.cc
-=======
   join_optimizer/hypergraph.cc
->>>>>>> mysql-server/8.0
+  ../libmysql/libmysql.cc
```
Comment (satya-bodapati) [inline sql/CMakeLists.txt]: According to upstream, libmysql.cc is not necessary. Is it required for PXB?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/CMakeLists.txt (hunk 14 above)
Comment (rahulmalik87) [reply sql/CMakeLists.txt]: it is there in the previous release as well. nothing changed in this release. was as part of PXB commit 6577bdfa719234315a5bc6d4f1042143582b2ef8. should I check further?

Comment (satya-bodapati) [reply sql/CMakeLists.txt]: Just remove and see if it compiles

**Comment by rahulmalik87** [comment]
all known bugs failures 
upstream 8.0.21 https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/354/ 

upstream 8.0.22 https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/353/

raised bug for https://jira.percona.com/browse/PXB-2358 version_check which is there is trunk as well. 
https://jira.percona.com/browse/PXB-2357 for redo_archive hang.

**Review by altmannmarcelo — APPROVED** [review]

File: sql/CMakeLists.txt (hunk 14 above)
Comment (rahulmalik87) [reply sql/CMakeLists.txt]: it is compiling on all environment.

Comment (satya-bodapati) [reply sql/CMakeLists.txt]: Please remove this then.

**Comment by satya-bodapati** [comment]
@rahulmalik87 Will you change XB_VERSION (to 8022) in this PR?
You answered yes, but didn't update this file. So double checking

File: sql/CMakeLists.txt (hunk 14 above)
Comment (rahulmalik87) [reply sql/CMakeLists.txt]: removing it. thanks for suggestion.

**Comment by rahulmalik87** [comment]
> @rahulmalik87 Will you change XB_VERSION (to 8022) in this PR?
> You answered yes, but didn't update this file. So double checking

it would be done as different PR, most probably by build team.

**Comment by altmannmarcelo** [comment]
> > @rahulmalik87 Will you change XB_VERSION (to 8022) in this PR?
> > You answered yes, but didn't update this file. So double checking
> 
> it would be done as different PR, most probably by build team.

@rahulmalik87 we do that as part of the merge. 

@altmannmarcelo For PXB, I see that it is done by build/release team.

File: cmake/mysql_version.cmake (hunk 8 above)
Comment (rahulmalik87) [reply cmake/mysql_version.cmake]: so ${file} is either XB_VERSION or MYSQL_VERSION

Comment (satya-bodapati) [reply cmake/mysql_version.cmake]: ok, lets leave it as is then

**Comment by satya-bodapati** [comment]
@rahulmalik87 , regarding Jenkins results:
1)

https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/353/


Resetting invalid page [page id: space=0, page number=176] type 20 to 13 when flushing.
2020-11-23T12:12:44.951873Z 1 [Note] [MY-011924] [InnoDB] Legacy atomic write buffer created
2020-11-23T12:12:44.952355Z 1 [Note] [MY-012923] [InnoDB] Creating shared tablespace for temporary tables
2020-11-23T12:12:44.952977Z 1 [Note] [MY-012265] [InnoDB] Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2020-11-23T12:12:45.022775Z 1 [Note] [MY-012266] [InnoDB] File './ibtmp1' size is now 12 MB.
2020-11-23T12:12:45.023718Z 1 [ERROR] [MY-012128] [InnoDB] posix_fallocate(): Failed to preallocate data for file ./#innodb_temp/temp_1.ibt, desired size 81920 Operating system error number 22. Check that the disk is not full or a disk quota exceeded. Make sure the file system supports this function. Refer to your operating system documentation for operating system error code information.
2020-11-23T12:12:45.024152Z 1 [ERROR] [MY-012929] [InnoDB] InnoDB Database creation was aborted with error Out of disk space. You may need to delete the ibdata1 file before trying to start up again.

***** Can you check with build team and resubmit again? They ran out of disk space

https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/354/#showFailuresLink

https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param-medium/354/BUILD_TYPE=debug,Host=min-focal-x64,xtrabackuptarget=xtradb80/testReport/junit/(root)/t_pxb-1824/sh/

**** What about pxb-1824 test failure?

2) Please compare "SKIP_INSTALL" from previous release and this release. These "SKIP_INSTALL" will skip creating un-necessary binaries for PXB repo. Ask @altmannmarcelo  about this ;)

**Comment by satya-bodapati** [comment]
```
cat VERSION 
MYSQL_VERSION_MAJOR=8
MYSQL_VERSION_MINOR=0
MYSQL_VERSION_PATCH=21
MYSQL_VERSION_EXTRA=
satya@satya-Lenovo-YOGA-920-13IKB-Glass:~/WORK/percona-xtrabackup$ rgrep "SKIP_INSTALL" .|wc -l
64
-----
cat MYSQL_VERSION 
MYSQL_VERSION_MAJOR=8
MYSQL_VERSION_MINOR=0
MYSQL_VERSION_PATCH=22
MYSQL_VERSION_EXTRA=
satya.bodapati@qa-sm-twina02:~/percona-xtrabckup-rahul$ rgrep "SKIP_INSTALL" .|wc -l
64
```
Assuming no new binaries, SKIP_INSTALL count seems OK

**Review by satya-bodapati — APPROVED** [review]
pxb-1824.sh failure has to be analyzed separately (if it is existing failure)
